utils/frame/remote-externalities/src/lib.rs

```diff
@@ -230,33 +266,99 @@ impl<B: BlockT> Builder<B> {
 }
 
 // RPC methods
-impl<B: BlockT> Builder<B> {
+impl<B: BlockT> Builder<'_, B> {
 	async fn rpc_get_head(&self) -> Result<B::Hash, &'static str> {
 		trace!(target: LOG_TARGET, "rpc: finalized_head");
-		RpcApi::<B>::finalized_head(&self.as_online().rpc()).await.map_err(|e| {
+		RpcApi::<B>::finalized_head(self.as_online().rpc_client()).await.map_err(|e| {
 			error!("Error = {:?}", e);
 			"rpc finalized_head failed."
 		})
 	}
 
-	/// Relay the request to `state_getPairs` rpc endpoint.
+	/// Get all the keys at `prefix` at `hash` using the paged, safe RPC methods.
+	async fn get_keys_paged(
+		&self,
+		prefix: StorageKey,
+		hash: B::Hash,
+	) -> Result<Vec<StorageKey>, &'static str> {
+		const PAGE: u32 = 512;
+		let mut last_key: Option<StorageKey> = None;
+		let mut all_keys: Vec<StorageKey> = vec![];
+		let keys = loop {
+			let page = RpcApi::<B>::get_keys_paged(
+				self.as_online().rpc_client(),
+				Some(prefix.clone()),
+				PAGE,
+				last_key.clone(),
+				Some(hash),
+			)
+			.await
+			.map_err(|e| {
+				error!(target: LOG_TARGET, "Error = {:?}", e);
+				"rpc get_keys failed"
+			})?;
+			all_keys.extend(page.clone());
+
+			if page.len() < (PAGE as usize) {
+				debug!(target: LOG_TARGET, "last page received: {}", page.len());
+				break all_keys;
+			} else {
+				let new_last_key =
+					all_keys.last().expect("all_keys is populated; has .last(); qed");
+				debug!(
+					target: LOG_TARGET,
+					"new total = {}, full page received: {:?}",
+					all_keys.len(),
+					HexDisplay::from(new_last_key)
+				);
+				last_key = Some(new_last_key.clone());
+			}
+		};
+
+		Ok(keys)
+	}
+
+	/// Synonym of `rpc_get_pairs_unsafe` that uses paged queries to first get the keys, and then
+	/// map them to values one by one.
 	///
-	/// Note that this is an unsafe RPC.
-	async fn rpc_get_pairs(
+	/// This can work with public nodes. But, expect it to be darn slow.
```

Comment:
tested on a synced "public-node"?

Reply:
It takes a minute or two, but the slowness is mostly a function of how big the state is. My comment is a bit exaggerated and silly tbh, it is slightly slower than the previous http model, and substantially more stable since it is paged.